tests/test_cli_misc.py

```diff
@@ -269,6 +272,106 @@ def test_error_no_global_config_file(self):
                          'command cowardly declines to create it.')
 
 
+class LoginTests(unittest.TestCase):
+    """Establish that the `tower-cli login` command works in the way
+    that we expect.
+    """
+    def setUp(self):
+        self.runner = CliRunner()
+
+    def test_no_arguments(self):
+        """
+        Establish that if `tower-cli login` is called with no arguments, it
+        complains
+        """
+        # Invoke the command.
+        result = self.runner.invoke(login)
+
+        # Ensure that we got a non-zero exit status
+        self.assertNotEqual(result.exit_code, 0)
+        self.assertIn('Missing argument "username"', result.output)
+
+    def test_oauth_unsupported(self):
+        """Establish that if `tower-cli login` is used on a Tower that
+        doesn't support OAuth2, it shows a meaningful error
+        """
+        with client.test_mode as t:
+            # You have to modify this internal private registry to
+            # register a URL endpoint that _doesn't_ have the version
+            # prefix
+            prefix = Client().get_prefix(include_version=False)
+            t._registry[six.text_type('{}o/').format(prefix)] = Resp(
+                '', 404, {}
+            )
+            result = self.runner.invoke(
+                login, ['bob', '--password', 'secret']
+            )
+
+        # Ensure that we got a non-zero exit status
+        self.assertNotEqual(result.exit_code, 0)
+        self.assertIn(
+            'Error: This version of Tower does not support OAuth2.0',
+            result.output
+        )
+
+    def test_personal_access_token(self):
+        """Establish that if `tower-cli login` is called with a username and
+        password, we obtain and write an oauth token to the config file
+        """
+        # Invoke the command.
+        mock_open = mock.mock_open()
+        with mock.patch('tower_cli.cli.misc.open', mock_open,
+                        create=True):
+            with mock.patch.object(os, 'chmod'):
+                with client.test_mode as t:
+                    # You have to modify this internal private registry to
+                    # register a URL endpoint that _doesn't_ have the version
+                    # prefix
+                    prefix = Client().get_prefix(include_version=False)
+                    t._registry[six.text_type('{}o/').format(prefix)] = Resp(
+                        '', 200, {}
+                    )
+                    t.register('/users/bob/personal_tokens/', json.dumps({
```

Comment:
I'm curious, is it ok to have 'bob' there instead of the user's pk?

Reply:
Named URLs :D